tracing/src/main/java/com/palantir/tracing/TraceLocal.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Provides trace local variables. Each trace gets its own independent copy of the variable.
+ *
+ * A trace local is either set (has a value) or unset (has no value).
+ *
+ * Outside of a trace (i.e. when {@link Tracer#hasTraceId()} is false) then the trace local is always unset.
+ */
+public final class TraceLocal<T> {
+
+    @Nullable
+    private final Function<? super TraceLocal<?>, T> initialValue;
+
+    private TraceLocal(@Nullable Supplier<T> initialValue) {
+        if (initialValue == null) {
+            this.initialValue = null;
+        } else {
+            // eagerly transform supplier to avoid allocation per invocation
+            // (computeIfAbsent takes a Function)
+            this.initialValue = _ignored -> initialValue.get();
```

Comment:
Let's add some nullness validation here -- always easier in an initial release than adding new validation later on.
```suggestion
            this.initialValue = _ignored -> Preconditions.checkNotNull(initialValue.get(), "TraceLocal initial value must not be null");
```